File: ext-php/examples/bench_replace_load.php
<?php
/**
 * Load bench matching examples/bench_replace_load.rs (large profile).
 *
 *   cargo build -p calc_spreadsheet_php --release
 *   php -d extension=./target/release/libcalc_spreadsheet_php.so \
 *       ext-php/examples/bench_replace_load.php
 */

const LEAF_COUNT = 1024;
const MID_WIDTH = 8192;
const MID_REFS = 8;
const HEAVY_WIDTH = 8192;
const HEAVY_REFS = 40;
const REPLACE_KEYS = 512;
const PLACEHOLDERS_PER_FORMULA = 6;
const WARMUP = 2;
const ITERATIONS = 5;

function buildSheetAndReplacements(): array
{
    $replacements = [];
    for ($i = 0; $i < REPLACE_KEYS; $i++) {
        $key = "__R{$i}__";
        if ($i % 17 === 0) {
            $replacements[$key] = "T{$i}";
        } else {
            $replacements[$key] = ($i % 97) + 1;
        }
    }

    $cells = [];
    for ($i = 0; $i < LEAF_COUNT; $i++) {
        $cells["L{$i}"] = (string) ($i % 97);
    }

    for ($i = 0; $i < MID_WIDTH; $i++) {
        $args = [];
        for ($j = 0; $j < MID_REFS; $j++) {
            $args[] = 'L' . (($i + $j) % LEAF_COUNT);
        }
        $ph = [];
        for ($j = 0; $j < PLACEHOLDERS_PER_FORMULA; $j++) {
            $k = ($i + $j * 3) % REPLACE_KEYS;
            if ($k % 17 === 0) {
                $k = ($k + 1) % REPLACE_KEYS;
            }
            $ph[] = "__R{$k}__";
        }
        $cells["M{$i}"] = '=SUM(' . implode(',', $args) . ',' . implode(',', $ph) . ')';
    }

    for ($i = 0; $i < 32; $i++) {
        $k = ($i * 17) % REPLACE_KEYS;
        $cells["C{$i}"] = "=\"id=\"&__R{$k}__&M{$i}";
    }

    for ($i = 0; $i < HEAVY_WIDTH; $i++) {
        $args = [];
        for ($j = 0; $j < HEAVY_REFS; $j++) {
            $args[] = 'M' . (($i + $j) % MID_WIDTH);
        }
        for ($j = 0; $j < PLACEHOLDERS_PER_FORMULA; $j++) {
            $k = ($i * 5 + $j) % REPLACE_KEYS;
            if ($k % 17 === 0) {
                $k = ($k + 1) % REPLACE_KEYS;
            }
            $args[] = "__R{$k}__";
        }
        $cells["H{$i}"] = '=SUM(' . implode(',', $args) . ')';
    }

    return [$cells, $replacements];
}

function measure(int $n, callable $fn): array
{
    $totalNs = 0;
    $minNs = PHP_INT_MAX;
    $last = null;
    for ($i = 0; $i < $n; $i++) {
        $t0 = hrtime(true);
        $last = $fn();
        $elapsed = hrtime(true) - $t0;
        $totalNs += $elapsed;
        if ($elapsed < $minNs) {
            $minNs = $elapsed;
        }
    }
    return [
        'avg_ms' => ($totalNs / $n) / 1e6,
        'min_ms' => $minNs / 1e6,
        'total_ms' => $totalNs / 1e6,
        'last' => $last,
    ];
}

echo "=== bench_replace_load (PHP) ===" . PHP_EOL;
echo 'PHP ' . PHP_VERSION . PHP_EOL;

$buildStart = hrtime(true);
[$cells, $replacements] = buildSheetAndReplacements();
$buildMs = (hrtime(true) - $buildStart) / 1e6;

$placeholderHits = 0;
foreach ($cells as $expr) {
    $placeholderHits += substr_count($expr, '__R');
}

echo 'cells in:           ' . count($cells) . PHP_EOL;
echo 'replacement keys:   ' . count($replacements) . PHP_EOL;
echo "placeholder tokens: ~{$placeholderHits}" . PHP_EOL;
echo sprintf("sheet build:        %.3f ms%s", $buildMs, PHP_EOL);
echo 'warmup=' . WARMUP . ', iterations=' . ITERATIONS . PHP_EOL;
echo PHP_EOL;

for ($i = 0; $i < WARMUP; $i++) {
    calc_spreadsheet($cells, $replacements);
    calc_spreadsheet($cells, []);
}

$with = measure(ITERATIONS, static fn () => calc_spreadsheet($cells, $replacements));
$without = measure(ITERATIONS, static fn () => calc_spreadsheet($cells, []));

$sample = $with['last']['H0'] ?? null;
if (!is_float($sample) && !is_int($sample)) {
    fwrite(STDERR, "expected numeric H0, got: " . var_export($sample, true) . PHP_EOL);
    exit(1);
}

echo "--- with replacements ---" . PHP_EOL;
echo sprintf(
    "  avg %.3f ms | min %.3f ms | total %.3f ms (n=%d)%s",
    $with['avg_ms'],
    $with['min_ms'],
    $with['total_ms'],
    ITERATIONS,
    PHP_EOL
);
echo '  cells out: ' . count($with['last']) . PHP_EOL;
echo '  sample H0 = ' . var_export($sample, true) . PHP_EOL;
echo PHP_EOL;

echo "--- without replacements ---" . PHP_EOL;
echo sprintf(
    "  avg %.3f ms | min %.3f ms | total %.3f ms (n=%d)%s",
    $without['avg_ms'],
    $without['min_ms'],
    $without['total_ms'],
    ITERATIONS,
    PHP_EOL
);
echo PHP_EOL;

$ratio = $with['avg_ms'] / max($without['avg_ms'], 1e-9);
echo sprintf(
    "replace path / empty path (avg): %.2fx (%.1f ms vs %.1f ms)%s",
    $ratio,
    $with['avg_ms'],
    $without['avg_ms'],
    PHP_EOL
);

File: examples/bench_replace_load.rs
//! Load test: large sheet with many `__PLACEHOLDER__` tokens + replacement map.
//!
//! ```text
//! cargo run --release --example bench_replace_load
//! ```

use std::collections::HashMap;
use std::time::{Duration, Instant};

use calc_spreadsheet::{
    calculate_spreadsheet, CalculateOptions, CellValue, ReplacementValue,
};

/// Leaf constants (no placeholders).
const LEAF_COUNT: usize = 1024;
/// Independent mid layer (each cell refs a few leaves + placeholders).
const MID_WIDTH: usize = 8192;
const MID_REFS: usize = 8;
/// Heavy dependent layer (wide + many refs) — may hit adaptive parallel.
const HEAVY_WIDTH: usize = 8192;
const HEAVY_REFS: usize = 40;
/// How many distinct `__Rnnn__` keys in the replacement map.
const REPLACE_KEYS: usize = 512;
/// Placeholders embedded per mid/heavy formula (cycled from the map).
const PLACEHOLDERS_PER_FORMULA: usize = 6;

const WARMUP: usize = 2;
const ITERATIONS: usize = 5;

fn main() {
    let (owned, replacements) = build_sheet_with_replacements();
    let cells: Vec<(&str, &str)> = owned
        .iter()
        .map(|(n, e)| (n.as_str(), e.as_str()))
        .collect();

    let placeholder_hits: usize = owned
        .iter()
        .map(|(_, e)| e.matches("__R").count())
        .sum();

    println!("=== bench_replace_load ===");
    println!("cells in:           {}", cells.len());
    println!("replacement keys:   {}", replacements.len());
    println!("placeholder tokens: ~{placeholder_hits} (in formulas before prepare)");
    println!("warmup={WARMUP}, iterations={ITERATIONS}");
    println!();

    let opts_with = CalculateOptions {
        replacements: Some(&replacements),
        ..Default::default()
    };
    let opts_empty = CalculateOptions::default();

    for _ in 0..WARMUP {
        let _ = calculate_spreadsheet(&cells, opts_with).unwrap();
        let _ = calculate_spreadsheet(&cells, opts_empty).unwrap();
    }

    let with_rep = measure(ITERATIONS, || {
        calculate_spreadsheet(&cells, opts_with).unwrap()
    });
    let no_rep = measure(ITERATIONS, || {
        calculate_spreadsheet(&cells, opts_empty).unwrap()
    });

    let sample = with_rep.last_outcome.as_ref().unwrap();
    let sample_key = format!("H0");
    let sample_val = sample.values.get(&sample_key).cloned();
    assert!(
        matches!(sample_val, Some(CellValue::Number(_))),
        "expected numeric H0, got {sample_val:?}"
    );
    assert!(sample.ignored_replacement_keys.is_empty());
    assert_eq!(sample.values.len(), no_rep.last_outcome.as_ref().unwrap().values.len());

    println!("--- with replacements (full path) ---");
    print_stats(&with_rep.stats);
    println!("  cells out: {}", sample.values.len());
    println!(
        "  sample {sample_key} = {:?}",
        sample_val.and_then(|v| v.as_number())
    );
    println!();
    println!("--- without replacements (same sheet; leftover __R*__ → 0 like blank) ---");
    print_stats(&no_rep.stats);
    println!();
    let overhead = with_rep.stats.avg.as_secs_f64() / no_rep.stats.avg.as_secs_f64();
    println!(
        "replace path / empty path (avg): {overhead:.2}x ({:.1} ms vs {:.1} ms)",
        with_rep.stats.avg.as_secs_f64() * 1000.0,
        no_rep.stats.avg.as_secs_f64() * 1000.0
    );
}

struct Stats {
    total: Duration,
    avg: Duration,
    min: Duration,
}

struct RunResult {
    stats: Stats,
    last_outcome: Option<calc_spreadsheet::SpreadsheetOutcome>,
}

fn measure<F>(iterations: usize, mut run: F) -> RunResult
where
    F: FnMut() -> calc_spreadsheet::SpreadsheetOutcome,
{
    let mut total = Duration::ZERO;
    let mut min = Duration::MAX;
    let mut last = None;
    for _ in 0..iterations {
        let started = Instant::now();
        let outcome = run();
        let elapsed = started.elapsed();
        total += elapsed;
        if elapsed < min {
            min = elapsed;
        }
        last = Some(outcome);
    }
    RunResult {
        stats: Stats {
            total,
            avg: total / iterations as u32,
            min,
        },
        last_outcome: last,
    }
}

fn print_stats(s: &Stats) {
    println!(
        "  avg {:.3} ms | min {:.3} ms | total {:.3} ms (n={ITERATIONS})",
        s.avg.as_secs_f64() * 1000.0,
        s.min.as_secs_f64() * 1000.0,
        s.total.as_secs_f64() * 1000.0
    );
}

fn build_sheet_with_replacements() -> (Vec<(String, String)>, HashMap<String, ReplacementValue>) {
    let mut replacements = HashMap::with_capacity(REPLACE_KEYS);
    for i in 0..REPLACE_KEYS {
        let key = format!("__R{i}__");
        // Mix numbers and a few text values used only in concat-style mid cells.
        if i % 17 == 0 {
            replacements.insert(key, ReplacementValue::from_text(format!("T{i}")));
        } else {
            replacements.insert(key, ReplacementValue::from_i64((i % 97) as i64 + 1));
        }
    }

    let mut cells = Vec::with_capacity(LEAF_COUNT + MID_WIDTH + HEAVY_WIDTH);

    for i in 0..LEAF_COUNT {
        cells.push((format!("L{i}"), (i % 97).to_string()));
    }

    // Mid: numeric SUM of leaves + product of placeholder numbers (skip text keys).
    for i in 0..MID_WIDTH {
        let mut args = Vec::with_capacity(MID_REFS + PLACEHOLDERS_PER_FORMULA);
        for j in 0..MID_REFS {
            args.push(format!("L{}", (i + j) % LEAF_COUNT));
        }
        let mut ph_sum = Vec::new();
        for j in 0..PLACEHOLDERS_PER_FORMULA {
            let k = (i + j * 3) % REPLACE_KEYS;
            // Prefer numeric keys (skip those that are text every 17).
            let k = if k % 17 == 0 { (k + 1) % REPLACE_KEYS } else { k };
            ph_sum.push(format!("__R{k}__"));
        }
        let expr = format!(
            "=SUM({},{})",
            args.join(","),
            ph_sum.join(",")
        );
        cells.push((format!("M{i}"), expr));
    }

    // A few concat cells to exercise & + text replacements.
    for i in 0..32 {
        let k = (i * 17) % REPLACE_KEYS;
        cells.push((
            format!("C{i}"),
            format!("=\"id=\"&__R{k}__&M{i}"),
        ));
    }

    // Heavy layer over mid cells + more placeholders.
    for i in 0..HEAVY_WIDTH {
        let mut args = Vec::with_capacity(HEAVY_REFS + PLACEHOLDERS_PER_FORMULA);
        for j in 0..HEAVY_REFS {
            args.push(format!("M{}", (i + j) % MID_WIDTH));
        }
        for j in 0..PLACEHOLDERS_PER_FORMULA {
            let k = (i * 5 + j) % REPLACE_KEYS;
            let k = if k % 17 == 0 { (k + 1) % REPLACE_KEYS } else { k };
            args.push(format!("__R{k}__"));
        }
        cells.push((format!("H{i}"), format!("=SUM({})", args.join(","))));
    }

    (cells, replacements)
}
